cmd/soroban-rpc/internal/db/ledgerentry.go

```diff
@@ -140,15 +140,13 @@ type ledgerEntryReadTx struct {
 	stmtCache              *sq.StmtCache
 	latestLedgerSeqCache   uint32
 	ledgerEntryCacheReadTx *transactionalCacheReadTx
+	ledgerReader           LedgerReader
 	tx                     db.SessionInterface
 	buffer                 *xdr.EncodingBuffer
 }
 
 func (l *ledgerEntryReadTx) GetLatestLedgerSequence() (uint32, error) {
-	if l.latestLedgerSeqCache != 0 {
-		return l.latestLedgerSeqCache, nil
-	}
-	latestLedgerSeq, err := getLatestLedgerSequence(context.Background(), l.tx, l.globalCache)
+	latestLedgerSeq, err := getLatestLedgerSequence(context.Background(), l.ledgerReader, l.globalCache)
 	if err == nil {
 		l.latestLedgerSeqCache = latestLedgerSeq
 	}
```

Comment:
Sorry, I think there's a subtle bug here that I overlooked. Check out the comment here https://github.com/stellar/soroban-rpc/blob/f7ebe3ebccb47d02edc85946e63544282af6bb10/cmd/soroban-rpc/internal/db/ledgerentry.go#L348-L353

if we use the global cache we to get the latest ledger seq we might be getting a newer ledger than the one bound to the ledgerEntryReadTx.

I think there's also a pre-existing bug with `NewTx()`:

https://github.com/stellar/soroban-rpc/blob/f7ebe3ebccb47d02edc85946e63544282af6bb10/cmd/soroban-rpc/internal/db/ledgerentry.go#L372-L377

we should be locking before starting the transaction just like how it's done in `NewCachedTx()`. I wonder if we can get rid of `NewTx()` and instead always use  `NewCachedTx()`

@2opremio will be to confirm

Reply:
@tamirms Yes you are right here. But I think the change to add the cache read in `getLatestLedgerSequence` will still work for the reader function here - https://github.com/stellar/soroban-rpc/blob/f7ebe3ebccb47d02edc85946e63544282af6bb10/cmd/soroban-rpc/internal/db/ledgerentry.go#L341. But yes, the one bound to tx should still use the local cache value. @2opremio Any thoughts?

Reply:
> I wonder if we can get rid of `NewTx()` and instead always use `NewCachedTx()`
> 
> @2opremio will be to confirm

I took a look at `NewTx()` vs `NewCachedTx()`. There is an advantage to not using the cache when it's not needed (e.g. in `getLedgerEntries`, where the the entries will never be queried multiple times) the cache would just cause GC pressure but ... we should measure whether it's worth the extra complexity.



> we should be locking before starting the transaction just like how it's done in `NewCachedTx()`

The cache isn't being used by the transaction for reads (AFAICT it's only passed to update the latestLedgerEntry when it's not initialized) so I don't think we need to lock earlier. Do you have a specific race condition in mind?

Reply:
> Do you have a specific race condition in mind?

assume the read only transaction is created and then, right before we read lock the cache, the ingestion transaction on a separate goroutine commits and updates the latest ledger to n+1. After we acquire the read lock, we set `latestLedgerSeqCache` to n+1. However, since the read only transaction was created before the ingestion transaction was comitted, all the queries we do on that transaction will be consistent with ledger n instead of ledger n+1

Reply:
Ah, true, then we should lock while creating the transaction.